### ISSUE TITLE
tf.nn.isotonic_regression returns incorrect gradient for a pooled block

### ISSUE THREAD

**Issue by ALinrunrun** [tensorflow/tensorflow#127234]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0, also reproduced on tf 2.22.0-dev20260904

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

Python 3.13.5

### Current behavior?

`tf.nn.isotonic_regression` returns an incorrect first derivative for a stable pooled block.

The reproducer evaluates decreasing isotonic regression on `[4, t, 2]` with `t = 0.7`. For `0 < t < 2`, the last two entries form a stable pooled block with value `(t + 2) / 2`. Therefore, the selected output `y[1]` has derivative `1/2` throughout an open neighborhood.

The forward value is computed correctly as `1.35`, but reverse-mode autodiff returns `1.0` for the derivative with respect to `t`. The expected derivative is `0.5`.

Expected behavior?

The first derivative of the selected `tf.nn.isotonic_regression` output at `t = 0.7` should be close to `0.5`, not `1.0`.

### Standalone code to reproduce the issue

```shell
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["TF_NUM_INTRAOP_THREADS"] = "1"
os.environ["TF_NUM_INTEROP_THREADS"] = "1"

import tensorflow as tf

def target(t):
    x = tf.stack([
        tf.constant(4, dtype=t.dtype),
        t,
        tf.constant(2, dtype=t.dtype),
    ])
    y, segments = tf.nn.isotonic_regression(
        inputs=x,
        decreasing=True,
        axis=0,
    )
    return y[1]

x = tf.constant(0.7, dtype=tf.float64)

with tf.GradientTape() as tape:
    tape.watch(x)
    y = target(x)

actual = tape.gradient(y, x)
expected = 0.5

print("forward:", y.numpy())
print("actual:", actual.numpy())
print("expected:", expected)

if abs(float(actual) - expected) > 1e-9:
    print("BUG REPRODUCED: tf.nn.isotonic_regression returns incorrect gradient for a pooled block")
else:
    print("not reproduced")
```

### Relevant log output

```shell
forward: 1.35
actual: 1.0
expected: 0.5
BUG REPRODUCED: tf.nn.isotonic_regression returns incorrect gradient for a pooled block
```

**Comment by Kayyuri** [comment]
Hi @ALinrunrun, Thank you for reporting this issue!

You've clearly demonstrated that the  `tf.nn.isotonic_regression` function appears to return an incorrect gradient for pooled blocks. The forward value is correct at `1.35`, but the gradient with respect to `t` is `1.0` instead of the expected `0.5`. This indicates an issue in the gradient computation for pooled blocks.

We’ll investigate the issue on our end and provide an update. Thanks!

**Comment by Venkat6871** [comment]
Hi @ALinrunrun, 
Thanks for providing the clear reproduction.
I tested the same reproduction on TensorFlow 2.21.0 and the latest `tf-nightly`, and I can reproduce the reported behavior in both [environments](https://colab.sandbox.google.com/gist/Venkat6871/ffd544a54eecc6953509953b1b3a7bab/127234_tf_2-21-0-nightly-v.ipynb).
The forward result is correct, but the gradient is `1.0` instead of the expected `0.5`.
I am escalating this for further investigation and a fix in the isotonic regression gradient path.
Thanks!

**Comment by Venkat6871** [comment]
Hi @ALinrunrun ,
Thank you for raising your concern here.
A fix for this issue has been proposed in this [PR1](https://github.com/tensorflow/tensorflow/pull/127550), [PR2](https://github.com/tensorflow/tensorflow/pull/127475). Once it is merged, this issue will be resolved.

Hi @Phoebus-Liu , @twelfthlabor,
Thank you for raising the PR to address this issue. We truly appreciate your initiative and the effort in contributing the fix.
